File: wiggle_subsequence_ruby/wiggle_subsequence.rb
class WiggleSubsequence
  class << self
    def wiggle_max_length(array)
      array.reduce([]) do |wiggle_sub, current|
        case wiggle_sub.length
        when 0; then wiggle_sub + [current]
        when 1
          if current != wiggle_sub[-1]
            wiggle_sub + [current]
          else
            wiggle_sub
          end
        else
          if ((current < wiggle_sub[-1]) && (wiggle_sub[-1] > wiggle_sub[-2])) || ((current > wiggle_sub[-1]) && (wiggle_sub[-1] < wiggle_sub[-2]))
            wiggle_sub + [current]
          elsif ((current < wiggle_sub[-1]) && (wiggle_sub[-1] < wiggle_sub[-2]))
            wiggle_sub[0..-2] + [current]
          elsif ((current > wiggle_sub[-1]) && (wiggle_sub[-1] > wiggle_sub[-2]))
            wiggle_sub[0..-2] + [current]
          else
            wiggle_sub
          end
        end
      end
        .length
    end
  end
end

File: wiggle_subsequence/lib/wiggle_subsequence.ex
defmodule WiggleSubsequence do
  @doc """
  ## Examples

  iex> WiggleSubsequence.calc([1, 7, 4, 9, 2, 5])
  6

  iex> WiggleSubsequence.calc([1, 17, 5, 10, 13, 15, 10, 5, 16, 8])
  7

  iex> WiggleSubsequence.calc([1, 2, 3, 4, 5, 6, 7, 8, 9])
  2
  """
  def calc(array) do
    array
    |> build_wiggle_sub()
    |> length()
  end

  defp build_wiggle_sub(array)  do
    array
    |> Enum.reduce([], &(append_or_replace_head(&1, &2)))
  end

  defp append_or_replace_head(a, []), do: [a]
  defp append_or_replace_head(a, [b]) when a != b, do: [a, b]
  defp append_or_replace_head(a, [b, c | rest]) when (a < b) and (b > c), do: [a, b, c | rest]
  defp append_or_replace_head(a, [b, c | rest]) when (a > b) and (b < c), do: [a, b, c | rest]
  defp append_or_replace_head(a, [b, c | rest]) when (a > b) and (b > c), do: [a, c | rest]
  defp append_or_replace_head(a, [b, c | rest]) when (a < b) and (b < c), do: [a, c | rest]
  defp append_or_replace_head(_, wiggle), do: wiggle
end
